pkg/controller/operators/labeller/labels.go

```diff
@@ -0,0 +1,153 @@
+package labeller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	operatorsv1 "github.com/operator-framework/api/pkg/operators/v1"
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	operatorsv1alpha2 "github.com/operator-framework/api/pkg/operators/v1alpha2"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/install"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/operators/decorators"
+)
+
+type ApplyConfig[T any] interface {
+	WithLabels(map[string]string) T
+}
+
+type Client[A ApplyConfig[A], T metav1.Object] interface {
+	Apply(ctx context.Context, cfg ApplyConfig[A], opts metav1.ApplyOptions) (result T, err error)
+}
+
+func hasLabel(obj metav1.Object) bool {
+	if labels := obj.GetLabels(); labels != nil {
+		if value, ok := labels[install.OLMManagedLabelKey]; ok && value == install.OLMManagedLabelValue {
+			return true
+		}
+	}
+	return false
+}
+
+func ObjectLabeler[T metav1.Object, A ApplyConfig[A]](
+	ctx context.Context,
+	logger *logrus.Logger,
+	check func(metav1.Object) bool,
+	applyConfigFor func(name, namespace string) A,
+	apply func(namespace string, ctx context.Context, cfg A, opts metav1.ApplyOptions) (T, error),
+) func(
+	obj interface{},
+) error {
+	return func(obj interface{}) error {
+		cast, ok := obj.(T)
+		if !ok {
```

Comment:
Would it be correct to assume that if we get a cast failure, it's entirely programmer error? And panicking would be better?

Reply:
I think I'd prefer a panic such that we fail as quickly as possible in a noticeable manner, vs. hoping someone notices a log message. WDYT?

Reply:
I don't have a strong opinion, just modeled this after what controllers normally do when e.g. casting the queue key to a string before parsing it as `namespace/name` - if any of that fails, it's definitely programmer error, but none of it panics

Reply:
Ok, let's not introduce panics to just a portion. Fine to leave as you have it. Thanks.